userapi/src/main/java/fk/prof/userapi/cache/ZKBasedCacheInfoRegistry.java

```diff
@@ -0,0 +1,264 @@
+import org.apache.curator.framework.recipes.locks.InterProcessMutex;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+
+import java.nio.charset.Charset;
+import java.time.LocalDateTime;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Shared storage based on zookeeper to store the mapping profile -> ip:port and maintain resource utilization counters
+ * for each node in cluster.
+ * It is used by {@link ClusterAwareCache}.
+ *
+ * zookeeper node structure:
+ * /fkprof-userapi                                                                    (session namespace)
+ * /fkprof-userapi/nodesInfo/{ip:port} -> {@link NodeLoadInfo}                                (ephemeral)
+ * /fkprof-userapi/profilesLoadStatus/{profileName} -> {@link ProfileResidencyInfo}           (ephemeral)
+ *
+ * For resource utilization counter, currently only cached profile count is stored
+ *
+ * @see NodeLoadInfo
+ * @see ProfileResidencyInfo
+ *
+ * Created by gaurav.ashok on 01/08/17.
+ */
+class ZKBasedCacheInfoRegistry implements CacheInfoRegistry {
+    private static final Logger logger = LoggerFactory.getLogger(ZKBasedCacheInfoRegistry.class);
+    private static final String distributedLockPath = "/global_mutex";
+
+    private final String zkNodesLoadInfoPath;
+    private final CuratorFramework curatorClient;
+    private final InterProcessLock interProcessLock;
+
+    private final AtomicReference<ConnectionState> connectionState;
+    private final AtomicReference<LocalDateTime> lastZkLostTime;
+    private final AtomicBoolean recentlyZkConnectionLost;
+
+    private final byte[] myResidencyInfoInBytes;
+    private final ProfileResidencyInfo myProfileResidencyInfo;
+    private Runnable cacheInvalidator;
+
+    private Counter lockAcquireTimeoutCounter = Util.counter("zk.lock.timeouts");
+    private Counter onReconnectFailures = Util.counter("zk.onreconnect.failures");
+
+    ZKBasedCacheInfoRegistry(CuratorFramework curatorClient, String myIp, int port, Runnable cacheInvalidator) {
+        this.curatorClient = curatorClient;
+
+        this.zkNodesLoadInfoPath = "/nodesInfo/" + myIp + ":" + port;
+        this.interProcessLock = new InterProcessMutex(curatorClient, distributedLockPath);
+        this.lastZkLostTime = new AtomicReference<>(LocalDateTime.MIN);
+        this.recentlyZkConnectionLost = new AtomicBoolean(false);
+        this.myProfileResidencyInfo = ProfileResidencyInfo.newBuilder().setIp(myIp).setPort(port).build();
+        this.myResidencyInfoInBytes = this.myProfileResidencyInfo.toByteArray();
+        this.curatorClient.getConnectionStateListenable().addListener(this::zkStateChangeListener,
+            Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setNameFormat("curator-state-listener").build()));
```

Comment:
I think the idea behind a single thread executor is to not have thread races in listener method, correct? Can we state the same in a comment here?